Poetry: support for publishing & viewing projects

breville opened the pull request:
With this change, the final level of `/s/poem-art` has two more buttons in the "Finish" dialog: "Add to Projects" and "Publish":

<img width="1280" alt="Screen Shot 2021-11-05 at 2 44 22 AM" src="https://user-images.githubusercontent.com/2205926/140446616-a50ae9bf-b44b-494c-8317-741e42f89e70.png">

The lower part of https://studio.code.org/projects/public also shows Poetry projects:

<img width="1009" alt="Screen Shot 2021-11-10 at 3 01 22 PM" src="https://user-images.githubusercontent.com/2205926/141047496-c1689501-643b-45c8-9d3f-aa2a26341961.png">

(The missing thumbnails are a known issue [tracked](https://codedotorg.atlassian.net/browse/LP-2093?atlOrigin=eyJpIjoiYjNiYjY5ZTVjMzQzNDhjZGExNWQ1ZmI0YmU1NWJkMDkiLCJwIjoiamlyYS1zbGFjay1pbnQifQ) separately.)

And the upper part of https://studio.code.org/projects/public shows a link to create a new Poetry project:

<img width="1016" alt="Screen Shot 2021-11-10 at 2 59 11 PM" src="https://user-images.githubusercontent.com/2205926/141047556-8987cc5a-3bb5-48b7-a396-91862749c9d1.png">

Erin007 commented on apps/src/p5lab/gamelab/GameLab.js:
```diff
@@ -19,6 +19,10 @@ export default class GameLab extends P5Lab {
+  saveToProjectGallery() {
+    return false;
+  }
+
```
Why do we need this and below for SpriteLab?  I think those project types can and should be publishable to the public gallery; if the saving to gallery here is a different gallery that I'm unaware of, can we specify which gallery in the function name please? Something like `saveToPersonalGallery` or `saveToSectionGallery`.

breville replied:
I'm definitely looking for guidance here.  My goal was to modify the sharing behaviour for the two new poetry-based tutorials without changing behaviour for any other levels in our system.  

If I understand things right, Poetry inherits from SpriteLab which inherits from P5Lab.  But I think there can also be other activities in our system which use SpriteLab and GameLab directly, and those are the ones I particularly didn't want to change in any way.

cc @ajpal

breville replied:
I've added some additional comments to the code to help explain what these are for.

maddiedierker replied:
it seems like this is the default behavior we want (saveToProjectGallery returns false), so this function should probably live in P5Lab.js and only labs like Poetry that want to enable this will need to override it

breville replied:
Good call.  Updated.

Erin007 reviewed (commented):
The work related to adding Poetry as a project type in the gallery looks good. I think you'll also want to add to projectConstants.js similar to https://github.com/code-dot-org/code-dot-org/pull/25459 so that Poetry shows up as a project type in the personal projects table and so that it's eligible to be featured. 

Can you share more details about what saving to gallery means? It's different than my understanding of how projects work today and I'll need a little more context to review fully. Thanks!

breville commented:
> The work related to adding Poetry as a project type in the gallery looks good. I think you'll also want to add to projectConstants.js similar to #25459 so that Poetry shows up as a project type in the personal projects table and so that it's eligible to be featured.

Good catch.  `PROJECT_TYPE_MAP` [already had _two_ entries](https://github.com/code-dot-org/code-dot-org/pull/43376/files#diff-5fcaa03a15e362fb8f725eb03acdf3feb0d6980ae9a0d212bf39da382f558b80R36-R37), and I [added one](https://github.com/code-dot-org/code-dot-org/pull/43376/files#diff-5fcaa03a15e362fb8f725eb03acdf3feb0d6980ae9a0d212bf39da382f558b80R62) to `FEATURED_PROJECT_TYPE_MAP`.

> Can you share more details about what saving to gallery means? It's different than my understanding of how projects work today and I'll need a little more context to review fully. Thanks!

I've updated the description with a screenshot of the "Finish" dialog and added some comments to the code that might help.  Much of the PR centers around this dialog, shown by P5Lab calling `this.studioApp_.displayFeedback`, and whether that call passes `saveToProjectGallery: true`, which adds the "Add to Projects" and "Publish" buttons.  The former appears to [save](https://github.com/code-dot-org/code-dot-org/blob/13563617bbc9c64a307ead937b0e96505b3f4b85/apps/src/feedback.js#L436-L455) the project, while the latter [publishes](https://github.com/code-dot-org/code-dot-org/blob/13563617bbc9c64a307ead937b0e96505b3f4b85/apps/src/feedback.js#L457-L513) it to the public gallery if the user confirms in this subsequent dialog:

<img width="794" alt="Screen Shot 2021-11-05 at 3 16 44 AM" src="https://user-images.githubusercontent.com/2205926/140464665-29542b9c-53db-44b7-ab0d-8c3c890cc6eb.png">

maddiedierker commented on apps/src/templates/projects/projectsRedux.js:
```diff
@@ -170,6 +171,7 @@ function projectLists(state = initialProjectListState, action) {
 const initialHasOlderProjects = {
```
i see the comment above this object, but i still don't really understand... do you know what this means?

Erin007 replied:
If I remember correctly, when we load the public gallery we only fetch 100 published projects of each type. `hasOlderProjects` determines whether there are more projects already on the client to show or more to fetch from the server, which in turn determines whether to hide or show the view more button for the project type. We have soooooooo many projects of most types though that there's a functionally endless supply so I don't know how critical this is.

Erin007 replied:
This has a little more context: https://github.com/code-dot-org/code-dot-org/pull/15631 (and interestingly, the feature idea for search in the public gallery)

Erin007 replied:
Also because these all default to `true` except `special_topic` which is Hackathon remnants of a feature we never fully implemented, it feels like (outside the context of this PR) we could definitely cleaning both `special_topic` and the `hasOlderProjects` and `initialHasOlderProjects` state up.

maddiedierker replied:
thank you, that explains it! i'm not sure which cabal would own it, but is there a ticket we can log (or maybe one already exists) for that cleanup?

Erin007 replied:
Logged in the LP backlog (https://codedotorg.atlassian.net/browse/LP-2107) but I can't say with any confidence we'll get to it any time soon.
